### PR TITLE
feat(ffi): Update IR stream protocol version handling in preparation for releasing the kv-pair IR stream format:

### PR DESCRIPTION
**Pull request by LinZhihao-723**
- Bump the IR stream protocol version to 0.1.0 for the kv-pair IR stream format.
- Treat the previous IR stream format's versions as backwards compatible.
- Differentiate between backwards-compatible and supported versions during validation.

# Description

This PR formally updates the version protocol to mark released versions as `0.1.0` for the key-value pair IR format.
Also, to support backward compatibility and not break the existing old IR format serialization/deserialization method, we add support for version validation for IR version `v0.0.0`, `0.0.1` and `0.0.2`, which is the unstructured IR format. It also updates the old serialization methods to stick to the latest compatible version (`0.0.2`) to differentiate with the key-value pair IR format.
This PR also introduces a minimum supported version to invalidate any version larger than `0.0.2` and smaller than `0.1.0`, since we don't want to support any compatibility to these versions we never formally released.
However, after realizing the current version validation function is broken since it's doing purely string comprising, we replace the validation function with hardcoded values instead. The proper implementation should be deferred to future PRs.

# Validation performed

- Add new unit tests to ensure the version validation works as expected.
- Ensure the existing unit tests all passed, meaning that both old IR serializer/deserializer and key-value pair IR serializer/deserializer still work with the correct version numbers.

## Summary by CodeRabbit

- **New Features**
	- Enhanced protocol version validation for improved error handling.
	- Updated metadata handling in serialization processes.
	- Introduced a strongly typed enum class for protocol error codes to enhance type safety.

- **Bug Fixes**
	- Streamlined error handling in deserialization processes for consistency.
	- Improved clarity in error reporting for protocol version validation.

- **Tests**
	- Refined test cases for protocol version validation and log event deserialization to ensure comprehensive coverage and clarity.

### CODE REVIEW

File: components/core/src/clp/ffi/ir_stream/decoding_methods.cpp [hunk 1]
```diff
@@ -468,33 +470,69 @@ IRErrorCode deserialize_preamble(
...
 
-IRProtocolErrorCode validate_protocol_version(std::string_view protocol_version) {
-    if ("v0.0.0" == protocol_version) {
-        // This version is hardcoded to support the oldest IR protocol version. When this version is
-        // no longer supported, this branch should be removed.
-        return IRProtocolErrorCode_Supported;
+auto validate_protocol_version(std::string_view protocol_version) -> IRProtocolErrorCode {
+    // These versions are hardcoded to support the IR protocol version that predates the key-value
+    // pair IR format.
+    constexpr std::array<std::string_view, 3> cBackwardCompatibleVersions{
```
Comment (junhaoliao) [inline components/core/src/clp/ffi/ir_stream/decoding_methods.cpp:476]: Do we want to include any backward compatible beta versions here?

Comment (LinZhihao-723) [reply components/core/src/clp/ffi/ir_stream/decoding_methods.cpp:476]: For old IR format no.
For key-value IR format, we do support `0.1.0-beta.1`. I added a util to strip the pre-release and only compare the version core.

File: components/core/src/clp/ffi/ir_stream/decoding_methods.hpp [hunk 2]
```diff
@@ -193,15 +194,15 @@ IRErrorCode deserialize_utc_offset_change(ReaderInterface& reader, UtcOffset& ut
...
  * @param protocol_version
- * @return IRProtocolErrorCode_Supported if the protocol version is supported.
- * @return IRProtocolErrorCode_Too_Old if the protocol version is no longer supported by this
- * build's protocol version.
- * @return IRProtocolErrorCode_Too_New if the protocol version is newer than this build's protocol
- * version.
- * @return IRProtocolErrorCode_Invalid if the protocol version does not follow the SemVer
+ * @return IRProtocolErrorCode::Supported if the protocol version is supported.
+ * @return IRProtocolErrorCode::BackwardCompatible if the protocol version indicates a stream format
+ * that predates the key-value pair IR format.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/ir_stream/decoding_methods.hpp:199]:
```suggestion
 * @return IRProtocolErrorCode::Supported if the protocol version is supported by the key-value
 * pair IR stream serializer and deserializer. TODO: Update this once we integrate backwards
 * compatibility into the deserializer.
 * @return IRProtocolErrorCode::BackwardCompatible if the protocol version is supported by the
 * serializer and deserializer for the IR stream format that predates the key-value pair IR stream
 * format. TODO: Update this once we integrate backwards compatibility into the key-value pair IR
 * stream format.
```

File: components/core/src/clp/ffi/ir_stream/protocol_constants.hpp [hunk 3]
```diff
@@ -12,8 +13,10 @@ constexpr int8_t LengthUByte = 0x11;
 constexpr int8_t LengthUShort = 0x12;
 
 constexpr char VersionKey[] = "VERSION";
-constexpr char VersionValue[] = "0.0.2";
-constexpr char BetaVersionValue[] = "0.1.0-beta.1";
+constexpr std::string_view VersionValue{"0.1.0"};
+constexpr std::string_view MinimumSupportedVersion{"0.1.0-beta.1"};
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/ir_stream/protocol_constants.hpp:17]: Do we need this? Do `0.1.0-beta.1` files exist in the wild, @junhaoliao?

File: components/core/src/clp/ffi/ir_stream/protocol_constants.hpp [hunk 4]
```diff
@@ -12,8 +13,10 @@ constexpr int8_t LengthUByte = 0x11;
...
+// This is used for IR stream format that predates the key-value pair IR format.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/ir_stream/protocol_constants.hpp:18]:
```suggestion

// This is used for the IR stream format that predates the key-value pair IR format.
```

Notice the preceding empty line before the comment; let's try to follow this practice to follow in Junhao's footsteps.

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 5]
```diff
@@ -844,17 +844,69 @@ TEST_CASE("decode_next_message_four_byte_timestamp_delta", "[ffi][deserialize_lo
...
+                std::string_view{"0.0.1"},
+                std::string_view{"0.0.2"}
+        )};
+        REQUIRE(
+                (clp::ffi::ir_stream::IRProtocolErrorCode::Backward_Compatible
+                 == validate_protocol_version(backward_compatible_versions))
+        );
+    }
+
+    SECTION("Test version too old") {
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:889]:
```suggestion
    SECTION("Test versions that're too old") {
```

File: components/core/tests/test-ir_encoding_methods.cpp [hunk 6]
```diff
@@ -844,17 +844,69 @@ TEST_CASE("decode_next_message_four_byte_timestamp_delta", "[ffi][deserialize_lo
...
+                std::string_view{"0.0.3-beta.1"},
+                std::string_view{"0.1.0-beta"}
+        )};
+        REQUIRE(
+                (clp::ffi::ir_stream::IRProtocolErrorCode::Unsupported
+                 == validate_protocol_version(old_versions))
+        );
+    }
+
+    SECTION("Test version too new") {
```
Comment (kirkrodrigues) [inline components/core/tests/test-ir_encoding_methods.cpp:901]:
```suggestion
    SECTION("Test versions that're too new") {
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/src/clp/ffi/ir_stream/decoding_methods.hpp [hunk 7]
```diff
@@ -20,12 +21,12 @@ typedef enum {
...
 
-typedef enum {
-    IRProtocolErrorCode_Supported,
-    IRProtocolErrorCode_Too_Old,
-    IRProtocolErrorCode_Too_New,
-    IRProtocolErrorCode_Invalid,
-} IRProtocolErrorCode;
+enum class IRProtocolErrorCode : uint8_t {
+    Supported,
+    Backward_Compatible,
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/ir_stream/decoding_methods.hpp:26]: hmmm, true, forgot to update the naming convention...

File: components/core/src/clp/ffi/ir_stream/protocol_constants.hpp (hunk 3 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/ir_stream/protocol_constants.hpp:17]: Talked with Junhao. I think we can drop support for `0.1.0-beta.1`

Comment (junhaoliao) [reply components/core/src/clp/ffi/ir_stream/protocol_constants.hpp:17]: Should be fine to skip. it's not yet deployed anywhere.

File: components/core/src/clp/ffi/ir_stream/decoding_methods.cpp [hunk 8]
```diff
@@ -468,33 +472,46 @@ IRErrorCode deserialize_preamble(
...
+            "v0.0.0",
+            "0.0.1",
+            cProtocol::Metadata::LatestBackwardCompatibleVersion
+    };
+    for (auto const backward_compatible_version : cBackwardCompatibleVersions) {
+        if (backward_compatible_version == protocol_version) {
+            return IRProtocolErrorCode::Backward_Compatible;
+        }
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/ir_stream/decoding_methods.cpp]: We shall defer this to future PRs.

**Review by kirkrodrigues — APPROVED** [review]
For the PR title & commit body, how about:

```
feat(ffi): Update IR stream protocol version handling in preparation for releasing the kv-pair IR stream format:
- Bump the IR stream protocol version to 0.1.0 for the kv-pair IR stream format.
- Treat the previous IR stream format's versions as backwards compatible.
- Differentiate between backwards-compatible and supported versions during validation.
```

**Pull request merged**